src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java

```diff
@@ -407,6 +412,13 @@ public void setLastModifiedTime(Path path, long newTime) throws IOException {
     }
   }
 
+  @Override
+  protected byte[] getFastDigest(Path path) throws IOException {
```

Comment:
I noticed that there's a discrepancy between what we do internally and what this PR is doing. Internally, the FileSystem implementation of getDigest() first calls getFastDigest(), and then, if that returns null, manually digests. Can you comment on why this shouldn't be called inside of getDigest()?

Reply:
This is related to this discussion we had earlier, right? https://github.com/bazelbuild/bazel/pull/11662#discussion_r449514054

The answer is that DigestUtils.getDigestWithManualFallback() already implements the right workflow for us. It digests can be obtained efficiently, it calls into getFastDigest(). When getFastDigest() doesn't return anything meaningful, it falls back to calling getDigest() (and caches the response in an apparently not-so-safe global cache).

Reply:
Yes, this is related to that discussion, sorry I didn't think it fully through then. I guess the possibilities are as follows:

(1) File is not an input to/output of any action. We never call getDigest() on it, only getFastDigest(), since we just compute a FileValue from it. (There is an exception: [RepositoryFunction](https://github.com/bazelbuild/bazel/blob/9ed89e5286be2f6cad0ff25ac442adeb84494faf/src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryFunction.java#L256).)
(2) File is an input/output to an action. Then we call getFastDigest() initially, and DigestUtils#getDigestWithManualFallback() before/after action execution.

It looks like we should never be calling Path#getDigest() at this point from any codepath other than DigestUtils. I wonder whether we should move DigestUtils to the vfs package and make Path#getDigest package-private.

For now, could you add a comment to Path#getDigest saying that it should not be called directly?

Reply:
That's a good idea. Done!